### PR TITLE
planner: correct row count estimation for inner child of index join

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

When there are `Selection` operators in inner child of IndexJoin, the row count for those operators is not that accurate, and it may lead to wrong join plan generated.

Take this case as an example:
```
mysql> explain analyze select count(*) from e, lo where lo.a=e.a and e.b=22336;
+----------------------------+----------+-----------+--------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| id                         | count    | task      | operator info                                                                  | execution info                                                                                                                       | memory                |
+----------------------------+----------+-----------+--------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| HashAgg_11                 | 1.00     | root      | funcs:count(1)                                                                 | time:1.229885346s, loops:2, rows:1                                                                                                   | N/A                   |
| └─IndexHashJoin_30         | 19977.00 | root      | inner join, inner:IndexLookUp_19, outer key:Column#4, inner key:Column#1       | time:1.229765777s, loops:23, rows:20044                                                                                              | 388.390625 KB         |
|   ├─TableReader_41         | 250.00   | root      | data:TableScan_40                                                              | time:8.33694ms, loops:6, rows:250, rpc time:8.31294ms                                                                                | 2.625 KB              |
|   │ └─TableScan_40         | 250.00   | cop[tikv] | table:lo, range:[-inf,+inf], keep order:false                                  | time:2ms, loops:4, rows:250                                                                                                          | N/A                   |
|   └─IndexLookUp_19         | 1.57     | root      |                                                                                | time:2.330220852s, loops:25, rows:20044, rpc max:486.462054ms, min:47.076912ms, avg:232.322106ms, p80:486.462054ms, p95:486.462054ms | 2.2626190185546875 MB |
|     ├─Selection_17         | 79.91    | cop[tikv] | not(isnull(Column#1))                                                          | proc max:422ms, min:44ms, p80:422ms, p95:422ms, rows:1020000, iters:1013, tasks:4                                                    | N/A                   |
|     │ └─IndexScan_15       | 79.91    | cop[tikv] | table:e, index:a, range: decided by [eq(Column#1, Column#4)], keep order:false | proc max:403ms, min:41ms, p80:403ms, p95:403ms, rows:1020000, iters:1013, tasks:4                                                    | N/A                   |
|     └─Selection_18         | 1.57     | cop[tikv] | eq(Column#2, 22336)                                                            | proc max:107ms, min:1ms, p80:68ms, p95:80ms, rows:20044, iters:1535, tasks:121                                                       | N/A                   |
|       └─TableScan_16       | 79.91    | cop[tikv] | table:e, keep order:false                                                      | proc max:107ms, min:1ms, p80:67ms, p95:80ms, rows:1020000, iters:1535, tasks:121                                                     | N/A                   |
+----------------------------+----------+-----------+--------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+-----------------------+

mysql> explain analyze select  /*+ TIDB_HJ(e, lo) */  count(*)  from e, lo where lo.a=e.a and e.b=22336;
+----------------------------+----------+-----------+------------------------------------------------------------------+---------------------------------------------------------------------------+-----------------+
| id                         | count    | task      | operator info                                                    | execution info                                                            | memory          |
+----------------------------+----------+-----------+------------------------------------------------------------------+---------------------------------------------------------------------------+-----------------+
| HashAgg_10                 | 1.00     | root      | funcs:count(1)                                                   | time:43.562803ms, loops:2, rows:1                                         | N/A             |
| └─HashLeftJoin_39          | 19977.00 | root      | inner join, inner:TableReader_53, equal:[eq(Column#1, Column#4)] | time:43.36269ms, loops:21, rows:20044                                     | 8.234375 KB     |
|   ├─IndexLookUp_51         | 19977.00 | root      |                                                                  | time:36.232663ms, loops:21, rows:20044, rpc time:19.571423ms              | 733.078125 KB   |
|   │ ├─IndexScan_48         | 19977.00 | cop[tikv] | table:e, index:b, range:[22336,22336], keep order:false          | time:18ms, loops:24, rows:20044                                           | N/A             |
|   │ └─Selection_50         | 19977.00 | cop[tikv] | not(isnull(Column#1))                                            | proc max:11ms, min:1ms, p80:11ms, p95:11ms, rows:20044, iters:44, tasks:6 | N/A             |
|   │   └─TableScan_49       | 19977.00 | cop[tikv] | table:e, keep order:false                                        | proc max:11ms, min:1ms, p80:11ms, p95:11ms, rows:20044, iters:44, tasks:6 | N/A             |
|   └─TableReader_53         | 250.00   | root      | data:TableScan_52                                                | time:6.818272ms, loops:2, rows:250, rpc time:6.883348ms                   | 2.6259765625 KB |
|     └─TableScan_52         | 250.00   | cop[tikv] | table:lo, range:[-inf,+inf], keep order:false                    | time:6ms, loops:4, rows:250                                               | N/A             |
+----------------------------+----------+-----------+------------------------------------------------------------------+---------------------------------------------------------------------------+-----------------+
```
The HashJoin execution is much faster than IndexJoin. Notice that estimated row count of `IndexScan_15` and `TableScan_16` is far away from the real scanned count as shown in explain analyze, i.e, `1020000 / 250 = 4080` is much bigger than estimated `79.91`, so cost of IndexJoin is under-estimated then.

### What is changed and how it works?

We should compute `countAfterAccess` and `countAfterIndex` with `avgInnerRowCount / selectivity`, instead of using `avgInnerRowCount * selectivity` to compute row count of `Selection`, because `avgInnerRowCount` is computed from `count of join / outer count`, and count of join has already taken the filters on `DataSource` into account in `LogicalJoin::DeriveStats`.

After this PR:
```
mysql> explain select count(*) from e, lo where lo.a=e.a and e.b=22336;
+--------------------------+----------+-----------+-------------------------------------------------------------------------------+
| id                       | count    | task      | operator info                                                                 |
+--------------------------+----------+-----------+-------------------------------------------------------------------------------+
| HashAgg_11               | 1.00     | root      | funcs:count(1)->Column#5                                                      |
| └─HashRightJoin_32       | 19977.00 | root      | inner join, inner:TableReader_41, equal:[eq(cost_model.lo.a, cost_model.e.a)] |
|   ├─TableReader_41       | 250.00   | root      | data:TableScan_40                                                             |
|   │ └─TableScan_40       | 250.00   | cop[tikv] | table:lo, range:[-inf,+inf], keep order:false                                 |
|   └─IndexLookUp_52       | 19977.00 | root      |                                                                               |
|     ├─IndexScan_49       | 19977.00 | cop[tikv] | table:e, index:b, range:[22336,22336], keep order:false                       |
|     └─Selection_51       | 19977.00 | cop[tikv] | not(isnull(cost_model.e.a))                                                   |
|       └─TableScan_50     | 19977.00 | cop[tikv] | table:e, keep order:false                                                     |
+--------------------------+----------+-----------+-------------------------------------------------------------------------------+
8 rows in set (0.01 sec)

mysql> explain select /*+ TIDB_INLJ(e) */ count(*) from e, lo where lo.a=e.a and e.b=22336;
+------------------------------+----------+-----------+---------------------------------------------------------------------------------------------+
| id                           | count    | task      | operator info                                                                               |
+------------------------------+----------+-----------+---------------------------------------------------------------------------------------------+
| HashAgg_10                   | 1.00     | root      | funcs:count(1)->Column#5                                                                    |
| └─IndexJoin_28               | 19977.00 | root      | inner join, inner:IndexLookUp_27, outer key:cost_model.lo.a, inner key:cost_model.e.a       |
|   ├─IndexLookUp_27           | 79.91    | root      |                                                                                             |
|   │ ├─Selection_25           | 4080.00  | cop[tikv] | not(isnull(cost_model.e.a))                                                                 |
|   │ │ └─IndexScan_23         | 4080.00  | cop[tikv] | table:e, index:a, range: decided by [eq(cost_model.e.a, cost_model.lo.a)], keep order:false |
|   │ └─Selection_26           | 79.91    | cop[tikv] | eq(cost_model.e.b, 22336)                                                                   |
|   │   └─TableScan_24         | 4080.00  | cop[tikv] | table:e, keep order:false                                                                   |
|   └─TableReader_40           | 250.00   | root      | data:TableScan_39                                                                           |
|     └─TableScan_39           | 250.00   | cop[tikv] | table:lo, range:[-inf,+inf], keep order:false                                               |
+------------------------------+----------+-----------+---------------------------------------------------------------------------------------------+
```
the row count estimation of `IndexScan_23` and `TableScan_24` is reasonable now, and we can make right decision for HashJoin then.

### Check List 

Tests 

 - Unit test
 - Integration test: this problem is found when adding integration tests to verify new cost model.

Code changes

N/A

Side effects

 - Possible performance regression: plan results of TPC-H have changed 
 - Increased code complexity

Related changes

N/A

Release note

N/A

### CODE REVIEW

**Comment by mahjonp** [comment]
/bench +tpch

```
workload:
  queries: 02.sql 19.sql
  repeat_time: 5
```

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -677,15 +677,26 @@ func (p *LogicalJoin) constructInnerTableScanTask(
...
+	selectivity := float64(1)
+	countAfterAccess := rowCount
+	if len(ts.filterCondition) > 0 {
+		var err error
+		selectivity, _, err = ds.tableStats.HistColl.Selectivity(ds.ctx, ts.filterCondition)
+		if err != nil || selectivity <= 0 {
+			logutil.BgLogger().Debug("unexpected selectivity, use selection factor")
+			selectivity = selectionFactor
+		}
+		countAfterAccess = rowCount / selectivity
```
Comment (lamxTyler) [inline planner/core/exhaust_physical_plans.go]: Add a upper bound for `countAfterAccess` and a comment for why divide selectivty here?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -769,26 +779,46 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
...
+		}
+		tmpPath.countAfterIndex = rowCount / selectivity
+		tmpPath.countAfterAccess = tmpPath.countAfterIndex
+	}
 	if len(indexConds) > 0 {
 		selectivity, _, err := ds.tableStats.HistColl.Selectivity(ds.ctx, indexConds)
-		if err != nil {
-			logutil.BgLogger().Debug("calculate selectivity failed, use selection factor", zap.Error(err))
+		if err != nil || selectivity <= 0 {
+			logutil.BgLogger().Debug("unexpected selectivity, use selection factor")
```
Comment (lzmhhh123) [inline planner/core/exhaust_physical_plans.go]: We'd better log more Infos such as selectivity value and table.

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -769,26 +779,46 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
...
 			selectivity = selectionFactor
 		}
-		tmpPath.countAfterIndex = rowCount * selectivity
+		tmpPath.countAfterAccess = tmpPath.countAfterIndex / selectivity
```
Comment (lzmhhh123) [inline planner/core/exhaust_physical_plans.go]: The name `countAfterAccess` is calculated under `indexConds`, that looks a little strange.

File: planner/core/exhaust_physical_plans.go (hunk 1 above)
Comment (eurekaka) [reply planner/core/exhaust_physical_plans.go]: `RowCount:     math.Min(1.0, countAfterAccess),` has set an upper bound, and for IndexScan, `ScaleByExpectCnt` can guarantee an upper bound.

File: planner/core/exhaust_physical_plans.go (hunk 3 above)
Comment (eurekaka) [reply planner/core/exhaust_physical_plans.go]: That is because we are backward reasoning from final row count.

**Pull request merged**